apps/arweave/src/ar_mining_cache.erl

```diff
@@ -185,50 +212,83 @@ with_cached_value(Key, SessionId, Cache0, Fun) ->
+				prometheus_gauge:inc(
+					mining_server_chunk_cache_release,
+					%% The actual amount of bytes reserved cannot go below zero, so we are
+					%% releasing all the reserved space at maximum.
+					min(Session#ar_mining_cache_session.reserved_mining_cache_bytes, -ReservationSizeAdjustment)
+				),
 				{ok, Session#ar_mining_cache_session{
 					reserved_mining_cache_bytes = max(0, Session#ar_mining_cache_session.reserved_mining_cache_bytes + ReservationSizeAdjustment)
 				}};
 			{ok, Value1} ->
-				SizeDiff = cached_size(Value1) - cached_size(Value0),
+				SizeDiff = cached_value_size(Value1) - cached_value_size(Value0),
 				SessionAvailableSize = available_size(Cache0) + Session#ar_mining_cache_session.reserved_mining_cache_bytes,
 				CacheLimit = get_limit(Cache0),
 				case SizeDiff > SessionAvailableSize of
 					true when CacheLimit =/= 0 -> {error, cache_limit_exceeded};
 					_ ->
+						%% Storing the cached value size of bytes (counting the diff between the old and new value).
+						prometheus_gauge:inc(
+							mining_server_chunk_cache_store, SizeDiff),
 						{ok, Session#ar_mining_cache_session{
 							mining_cache = maps:put(Key, Value1, Session#ar_mining_cache_session.mining_cache),
 							reserved_mining_cache_bytes = max(0, Session#ar_mining_cache_session.reserved_mining_cache_bytes - SizeDiff),
 							mining_cache_size_bytes = Session#ar_mining_cache_session.mining_cache_size_bytes + SizeDiff
 						}}
 				end;
 			{ok, Value1, ReservationSizeAdjustment} when ReservationSizeAdjustment < 0 ->
-				SizeDiff = cached_size(Value1) - cached_size(Value0) + ReservationSizeAdjustment,
+				SizeDiff = cached_value_size(Value1) - cached_value_size(Value0),
 				SessionAvailableSize = available_size(Cache0) + Session#ar_mining_cache_session.reserved_mining_cache_bytes,
 				CacheLimit = get_limit(Cache0),
 				case SizeDiff > SessionAvailableSize of
 					true when CacheLimit =/= 0 -> {error, cache_limit_exceeded};
 					_ ->
+						%% Storing the cached value size of bytes (counting the diff between the old and new value).
+						prometheus_gauge:inc(
+							mining_server_chunk_cache_store, SizeDiff),
+						%% Releasing the reserved space in the same amount of bytes, plus the reservation size adjustment.
+						prometheus_gauge:inc(
+							mining_server_chunk_cache_release,
+							%% The actual amount of bytes reserved cannot go below zero, so we are
+							%% releasing all the reserved space at maximum.
+							min(Session#ar_mining_cache_session.reserved_mining_cache_bytes, -ReservationSizeAdjustment)
+						),
 						{ok, Session#ar_mining_cache_session{
 							mining_cache = maps:put(Key, Value1, Session#ar_mining_cache_session.mining_cache),
-							reserved_mining_cache_bytes = max(0, Session#ar_mining_cache_session.reserved_mining_cache_bytes - SizeDiff),
+							reserved_mining_cache_bytes = max(0, Session#ar_mining_cache_session.reserved_mining_cache_bytes - SizeDiff + ReservationSizeAdjustment),
```

Comment:
Was this part of the issue? It looks like the old code did: `- (cached_value_size(Value1) - cached_value_size(Value0) + ReservationSizeAdjustment)` so that when expanded we ended up with: `-cached_value_size(Value1) + cached_value_size(Value0) - ReservationSizeAdjustment`

And the new code does `- ((cached_value_size(Value1) - cached_value_size(Value0)) + ReservationSizeAdjustment` so that when it expanded is: `-cached_value_size(Value1) + cached_value_size(Value0) + ReservationSizeAdjustment`

A difference of `- ReservationSizeAdjustment` vs. `+ ReservationSizeAdjustment`

Reply:
Yes, this was the very first issue fixed: the adjustment was not calculated correctly, which led to the actual cached size set wrong.
The last fix was about the reservation size though, the fix contained in the mining worker, not the cache.